sim/battle.ts

```diff
@@ -1771,6 +1819,110 @@ export class Battle extends Dex.ModdedDex {
+				if (effect.effectType === 'Weather' && !target.runStatusImmunity(effect.id)) {
+					this.debug('weather immunity');
+					retVals[i] = 0;
+					continue;
+				}
+				targetDamage = this.runEvent('Damage', target, source, effect, targetDamage);
+				if (!(targetDamage || targetDamage === 0)) {
+					this.debug('damage event failed');
+					retVals[i] = curDamage;
+					continue;
+				}
+			}
+			if (targetDamage !== 0) targetDamage = this.clampIntRange(targetDamage, 1);
+
+			if (this.gen <= 1) {
+				if (this.currentMod === 'stadium' ||
+					!['recoil', 'drain'].includes(effect.id) && effect.effectType !== 'Status') {
+					this.lastDamage = targetDamage;
+				}
+			}
+
+			retVals[i] = targetDamage = target.damage(targetDamage, source, effect);
+			if (targetDamage !== 0) target.hurtThisTurn = true;
+			if (source && effect.effectType === 'Move') source.lastDamage = targetDamage;
+
+			const name = effect.fullname === 'tox' ? 'psn' : effect.fullname;
+			switch (effect.id) {
+			case 'partiallytrapped':
+				this.add('-damage', target, target.getHealth, '[from] ' + this.effectData.sourceEffect.fullname, '[partiallytrapped]');
+				break;
+			case 'powder':
+				this.add('-damage', target, target.getHealth, '[silent]');
+				break;
+			case 'confused':
+				this.add('-damage', target, target.getHealth, '[from] confusion');
+				break;
+			default:
+				if (effect.effectType === 'Move' || !name) {
+					this.add('-damage', target, target.getHealth);
+				} else if (source && (source !== target || effect.effectType === 'Ability')) {
+					this.add('-damage', target, target.getHealth, '[from] ' + name, '[of] ' + source);
+				} else {
+					this.add('-damage', target, target.getHealth, '[from] ' + name);
+				}
+				break;
+			}
+
+			if (targetDamage) {
+				if (this.gen <= 1 && effect.recoil && source) {
+					this.damage(this.clampIntRange(Math.floor(targetDamage * effect.recoil[0] / effect.recoil[1]), 1), source, target, 'recoil');
+				}
+				if (this.gen <= 4 && effect.drain && source) {
+					this.heal(this.clampIntRange(Math.floor(targetDamage * effect.drain[0] / effect.drain[1]), 1), source, target, 'drain');
+				}
+				if (this.gen > 4 && effect.drain && source) {
+					this.heal(Math.round(targetDamage * effect.drain[0] / effect.drain[1]), source, target, 'drain');
+				}
+			}
+		}
+
+		// @ts-ignore TODO: AfterDamage passes an Effect, not an ActiveMove
+		if (!effect.flags) effect.flags = {};
+		if (instafaint) {
+			for (const [i, target] of targetArray.entries()) {
+				if (!retVals[i] || !target) continue;
+
+				if (target.hp <= 0) {
+					this.debug('instafaint: ' + this.faintQueue.map(entry => entry.target.name));
+					this.faintMessages(true);
+					if (this.gen <= 2) {
+						target.faint();
+						if (this.gen <= 1) this.queue = [];
+					}
+				}
+			}
+		}
+		retVals = this.runEvent('AfterDamage', (targetArray.filter(val => !!val)) as Pokemon[], source, effect, retVals);
+
+		return retVals;
+	}
 	damage(
```

Comment:
Can `damage` just delegate to `spreadMoveDamage`? That might make the diff clearer.

Reply:
Oh yeah, that's probably fine.